cloudsim-plus/src/main/java/org/cloudbus/cloudsim/brokers/DatacenterBrokerBestFit.java

```diff
@@ -0,0 +1,107 @@
+package org.cloudbus.cloudsim.brokers;
+
+import org.cloudbus.cloudsim.cloudlets.Cloudlet;
+import org.cloudbus.cloudsim.core.CloudSim;
+import org.cloudbus.cloudsim.core.events.SimEvent;
+import org.cloudbus.cloudsim.vms.Vm;
+
+import javax.swing.text.html.Option;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+
+/**
+ * <p>A simple implementation of {@link DatacenterBroker} that uses a best fit
+ * mapping among submitted cloudlets and Vm's.
+ * The Broker then places the submitted Vm's at the first Datacenter found.
+ * If there isn't capacity in that one, it will try the other ones.</p>
+ *
+ * @author Humaira Abdul Salam
+ */
+public class DatacenterBrokerBestFit extends DatacenterBrokerSimple {
+
+    Map<Long, Long> vmNumberOfPesUpdated;
```

Comment:
This `Map` was needed not only to calculate the free `Pe`'s, but also to keep track of the `Pe`'s that would be used at a later point in the simulation due to the `Vm`'s being mapped to the `Cloudlet`'s. I try to explain this further below using a control flow of a typical simulation:
- After the `DatacenterBroker` receives the acknowledgment of the creation of `Vm`'s, the broker maps the `Vm`'s to all `Cloudlet`'s in the `cloudletWaitingList` (see [this loop](https://github.com/humaira-salam/cloudsim-plus/blob/d0ff7dbda7f5e6d641e78bfe5c84f99205c508a6/cloudsim-plus/src/main/java/org/cloudbus/cloudsim/brokers/DatacenterBrokerAbstract.java#L887))
- After the `Vm` has been selected for a `Cloudlet` by the `vmMapper`, the `Vm` object is not updated  and the above mentioned loop just cycle through all the `Cloudlet`'s and keeps assuming throughout the execution of the loop that `Vm`'s resources are not changing. Although a smarter way for a broker would be to keep track of the resources reserved for a `Cloudlet` in each call of the `vmMapper`. This the reason I wanted to maintain a `Map` of `freePesNumber` in the broker.

With your suggestion of using a `freePesNumber` attribute, I will still not be able to solve the problem of tracking the resources being reserved during each call of the `vmMapper`. This will only update after `CLOUDLET_SUBMIT` messages are received by the `Datacenter`. But by that time, it is too late as the mapping is not done again (right?). 

Is it deliberately done so (mapping first and updating the `Vm` object later)? If yes, then another workaround would be to **not** submit 100's of `Cloudlet`'s at the beginning of the simulation but rather dynamically submit them during the simulation so that `Vm`'s can have chance to be updated. But somehow this doesn't seem like the best solution. Please let me know what your take is in this regard.

Reply:
I see your point and you're right. The cloudlet's processing is deliberately done only after Cloudlets are submitted because only after that, Cloudlets start to execute.
Just submitting Cloudlets during simulation runtime is not a reasonable solution. Lots of researchers submit most part of their Cloudlets before the simulation starts.

I think that a suitable solution would be to define an expectedUsedPesNumber attribute inside the VmSimple class. This attribute may be updated when a Cloudlet is linked/submitted to a VM and when a Cloudlet finishes executing. This update may follow the details I provided above in the previous comment.

Reply:
Following that approach, we'll have the expected number of used PEs, before the Cloudlets start to execute. This way, we don't need to use a Map.

Looking for a value inside a Map may be a really expensive operation as the Map grows.
For some experiments that I've created, all the performance issues that I've been fixing were causing the simulation to run for more than 5 hours, using more than 6GB of RAM and yet I couldn't see the end of the simulation because I just gave up waiting.

Reply:
So I added the expectedUsedPesNumber attribute. Thanks for the clue. Alright will try to avoid Maps as much as possible then!

Reply:
> Since we need to update the number of free PEs, the CloudletSchedulerAbstract.java already iterates over the list of running Cloudlets inside the [updateProcessing method](https://github.com/manoelcampos/cloudsim-plus/blob/master/cloudsim-plus/src/main/java/org/cloudbus/cloudsim/schedulers/cloudlet/CloudletSchedulerAbstract.java#L530).
> This way, we avoid re-iterating over that list again (that would impact the performance of large scale simulations).
> 
> However, there is a catch in how to determine the number of free PEs. We could simply sum the number of PEs from each running Cloudlet. But due to the way CloudletSchedulerTimeShared is implemented (see #33), the result may be incorrectly higher than the number of existing PEs.
> To address such an issue, you can implement the `setFreePessNumber(int freePesNumber)` inside VmSimple as below:
> 
> ```java
>   if(freePesNumber < 0){
>     throw new IllegalArgumentException("Number of free PEs cannot be negative.");
>   }
>   this.freePesNumber = Math.min(freePesNumber, getNumberOfPes());
> ```
> 
> This way, we don't need that Map, improving performance for large scale simulations.

Although for the best fit approach, I am using the expectedFreePesNumber attribute from Vm now, but I kept the freePesNumber attribute still in the class for any future use-case.

Also if freePesNumber is negative I put out a warning and set the freePesNumber to zero,  instead of throwing an exception. Because the freePesNumber would usually go negative when mapping using Simple/Heuristic DatacenterBroker.